src/expr/src/expr/expr_trim_array.rs

```diff
@@ -0,0 +1,99 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use risingwave_common::array::{ListRef, ListValue};
+use risingwave_common::types::ScalarRefImpl;
+use risingwave_expr_macro::function;
+
+use crate::error::ExprError;
+use crate::Result;
+
+/// Trims an array by removing the last n elements. If the array is multidimensional, only the first
+/// dimension is trimmed.
+///
+/// Examples:
+///
+/// ```slt
+/// query T
+/// select trim_array(array[1,2,3,4,5,null::int], 4);
+/// ----
+/// {1,2}
+///
+/// query T
+/// select trim_array(array[1,2,3,4,5,null::int], 0);
+/// ----
+/// {1,2,3,4,5,NULL}
+///
+/// query T
+/// select trim_array(array[1,2,3,4,5,null::int], null::int);
+/// ----
+/// NULL
+///
+/// query T
+/// select trim_array(array[1,2,3,4,5,null::int], 6);
+/// ----
+/// {}
+///
+/// query T
+/// select trim_array(null::int[], 1);
+/// ----
+/// NULL
+///
+/// statement error
+/// select trim_array(array[1,2,3,4,5,null::int], 7);
+///
+/// statement error
+/// select trim_array(array[1,2,3,4,5,null::int], -1);
+///
+/// statement error
+/// select trim_array(array[1,2,3,4,5,null::int], 3.14);
+///
+/// statement error
+/// select trim_array(array[1,2,3,4,5,null::int], null::int[]);
+///
+/// statement error
+/// select trim_array(array[1,2,3,4,5,null::int], true);
+/// ```
+#[function("trim_array(list, int32) -> list")]
+fn trim_array(array: Option<ListRef<'_>>, n: Option<i32>) -> Result<Option<ListValue>> {
```

Comment:
This is a [strict](https://www.postgresql.org/docs/15/catalog-pg-proc.html#:~:text=not%20leak%2Dproof.-,proisstrict,-bool) function and never returns null when neither input is null. So it could be:
```
fn trim_array(array: ListRef<'_>, n: i32) -> Result<ListValue>
```

FYI checking function strictness in PostgreSQL:
```
test=# select proname, proisstrict from pg_proc where proname in ('trim_array', 'array_positions');
     proname     | proisstrict 
-----------------+-------------
 array_positions | f
 trim_array      | t
(2 rows)
```